File: 90-subsets-ii/subsets-ii.cpp
class Solution {
    void f(vector<int>& num, int index, vector<int>& current, vector<vector<int>>& res){
        res.push_back(current);
        for(int i = index; i<num.size(); ++i){
            if(index != i && num[i] == num[i-1]) continue;
            current.push_back(num[i]);
            f(num, i+1,current, res);
            current.pop_back();
        }
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> res;
        vector<int> current;
        f(nums, 0, current, res);
        return res;

        
    }
};

File: 90-subsets-ii/subsets-ii.java
class Solution {
    public void f(int[] nums, List<List<Integer>> res, List<Integer> curr, int i){
        if(i == nums.length) return;
        int j = i;
        while( j < nums.length && nums[j] == nums[i]) j++;
        f(nums, res, curr, j);
        curr.add(nums[i]);
        res.add(new ArrayList<>(curr));
        f(nums, res, curr, i+1);
        curr.remove(curr.size()-1);
    
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> curr = new ArrayList<Integer>();
        result.add(curr);
        f(nums, result, curr, 0);
        return result;
        
    }
}